Swap UnrollAndJam Pass to before the SLP Vectorizer Pass

adprasad-nvidia opened the pull request:
This change moves the LoopUnrollAndJam pass before the SLPVectorizerPass.
This will, if LoopUnrollAndJam is enabled, enable some outer loop vectorization in LLVM. This will both improve runtimes and produce fewer lines of assembly for the same programs.
Note that if IsFullLTO is enabled, LoopUnrollAndJam already occurs before the SLPVectorizerPass, so only the code adding the pass when IsFullLTO is not enabled is moved to before the SLPVectorizerPass.
The change does not regress performance on TSVC-2, RAJAPerf, and Coremark.

adprasad-nvidia committed:
[UnJ] Move LoopUnrollAndJamPass before SLPVectorizerPass

adprasad-nvidia committed:
[UnJ] Add comments explaining new position of UnrollAndJam

nikic reviewed (commented):
Please add a test in llvm/test/Transforms/PhaseOrdering that shows why changing the pipeline position is useful.

tschuett commented:
Do you have numbers for the accelerated applications?

nikic commented:
Compile-time:
Enabling unroll&jam: http://llvm-compile-time-tracker.com/compare.php?from=0606c64da8b73768aded766b11e37fd77365e39d&to=a8ac00e6e982c9ecde796ebeb39e258fd12e759a&stat=instructions:u
Moving the position: http://llvm-compile-time-tracker.com/compare.php?from=a8ac00e6e982c9ecde796ebeb39e258fd12e759a&to=581c57bc463076c736abca8fcf16b74143c10d33&stat=instructions:u

As you can see, the new position is a lot more expensive (and would prevent future default enablement). Two things I notice:

 * With full LTO, you are now running unroll&jam twice -- I assume that's not intentional and you want a `!IsFullLTO` condition?
 * In the new position, we get less analysis reuse -- most importantly I expect that this recomputes SCEV. Possibly moving the pass to a different position before SLP would give better results (e.g. directly after LoopVectorize? I don't know.)

Can't really comment on the change itself, as I'm not familiar with the pass.

adprasad-nvidia committed:
[UnJ] Do not run UnrollAndJam twice if full LTO

sjoerdmeijer commented:
> Can't really comment on the change itself, as I'm not familiar with the pass.

Performance numbers shows that this doesn't regress existing benchmarks, and by moving the pass we enable more vectorisation. I think that makes this a very sensible change. 

@nikic 's suggestions are good things to fix. About compile-times and reuse of analysis results, I was curious if that is just a bit of trial-and-error, or is there a smarter way to see if a pass is recalculating its required analysis passes? Can this e.g. be dumped in debug logs?

yashssh commented:
We can compare the output of `-debug-pass-manager` to see which analysis were invalidated and which were re-run. https://godbolt.org/z/ch8vWhsYx

yashssh commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -1306,6 +1307,12 @@ void PassBuilder::addVectorPasses(OptimizationLevel Level,
+        LoopUnrollAndJamPass(Level.getSpeedupLevel())));
+  }
   // Optimize parallel scalar instruction chains into SIMD instructions.
   if (PTO.SLPVectorization) {
```
Is SLP vectorization on be default? If not then instead of moving UnJ, running another instance of the pass under this condition can be explored to avoid compile time penalties?

adprasad-nvidia replied:
Yeah, SLP is on by default, so unfortunately this wouldn't work because of the compile time penalty

adprasad-nvidia committed:
[UnJ] Add test in Transforms/PhaseOrdering for outer loop vectorization

adprasad-nvidia committed:
[UnJ] Run UnJ with !IsFullLTO in same place as UnJ with IsFullLTO

adprasad-nvidia committed:
[UnJ] [SimplifyCFG] Only run extra SimplifyCFGPass if UnJ enabled

adprasad-nvidia commented:
@nikic:
- Running UnrollAndJam was a mistake, I have fixed that. 
- I have added a test in `llvm/test/Transforms/PhaseOrdering/outer-loop-vectorize.ll` that demonstrates how the new code can generate vectorized code from an outer loop IR. The same test with UnrollAndJam in its current position will fail as scalar code is generated.
- I have also followed your suggestion of moving UnrollAndJam earlier to enable greater analysis re-useby moving it to directly after LoopVectorize (after the InferAlignment pass). UnrollAndJam under `!IsFullLTO` now runs in the same place as UnrollAndJam under `IsFullLTO` (Doing this also required another SimplifyCFG pass to be run before UnrollAndJam for outer loop vectorization to trigger.) **This seems to fix the compile time regressions versus ToT with unroll-and-jam enabled.**

@tschuett: As @sjoerdmeijer said, the numbers do not regress existing benchmarks. The new test demonstrates that the code is being vectorized.

adprasad-nvidia committed:
[UnJ] [SimplifyCFG] Fix comment formatting

yashssh commented on llvm/test/Transforms/PhaseOrdering/outer-loop-vectorize.ll:
```diff
@@ -0,0 +1,174 @@
+; RUN: opt -passes='default<O3>' -enable-unroll-and-jam -allow-unroll-and-jam -S < %s | FileCheck %s
+
+target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32"
+target triple = "aarch64-unknown-linux-gnu"
```
Remove triple

yashssh commented on llvm/test/Transforms/PhaseOrdering/outer-loop-vectorize.ll:
```diff
@@ -0,0 +1,174 @@
+
+declare i32 @dummy(ptr noundef, ptr noundef, ptr noundef, ptr noundef, ptr noundef, ptr noundef, ptr noundef, ptr noundef, float noundef nofpclass(nan inf)) local_unnamed_addr #1
+
+attributes #0 = { nounwind uwtable vscale_range(1,16) "approx-func-fp-math"="true" "frame-pointer"="non-leaf" "no-infs-fp-math"="true" "no-nans-fp-math"="true" "no-signed-zeros-fp-math"="true" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="neoverse-v2" "target-features"="+bf16,+complxnum,+crc,+dotprod,+fp-armv8,+fp16fml,+fullfp16,+i8mm,+jsconv,+lse,+mte,+neon,+outline-a ...
+attributes #1 = { "approx-func-fp-math"="true" "frame-pointer"="non-leaf" "no-infs-fp-math"="true" "no-nans-fp-math"="true" "no-signed-zeros-fp-math"="true" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="neoverse-v2" "target-features"="+bf16,+complxnum,+crc,+dotprod,+fp-armv8,+fp16fml,+fullfp16,+i8mm,+jsconv,+lse,+mte,+neon,+outline-atomics,+pauth,+rand,+ras,+rcpc,+rdm, ...
+attributes #2 = { nounwind }
+
+!llvm.module.flags = !{!0, !1, !2, !3, !4}
+!llvm.ident = !{!5}
+
+!0 = !{i32 1, !"wchar_size", i32 4}
+!1 = !{i32 8, !"PIC Level", i32 2}
+!2 = !{i32 7, !"PIE Level", i32 2}
+!3 = !{i32 7, !"uwtable", i32 2}
+!4 = !{i32 7, !"frame-pointer", i32 1}
+!5 = !{!"clang version 19.0.0git (git@github.com:sjoerdmeijer/llvm-project.git 6efcff18dfc42038bafa67091e990b9c1b839a71)"}
+!6 = distinct !{!6, !7}
+!7 = !{!"llvm.loop.mustprogress"}
+!8 = !{!9, !9, i64 0}
+!9 = !{!"float", !10, i64 0}
+!10 = !{!"omnipotent char", !11, i64 0}
+!11 = !{!"Simple C/C++ TBAA"}
+!12 = distinct !{!12, !7, !13}
+!13 = !{!"llvm.loop.unroll_and_jam.count", i32 4}
+!14 = distinct !{!14}
```
Can simplify the test by getting rid of unnecessary attributes and metadata

yashssh reviewed (commented):
A pass ordering test would be nice but I'm not sure where that might go.
